### PR TITLE
ARROW-6798: [CI] [Rust] Build image contains cached dependencies

### PR DESCRIPTION
**Pull request by andygrove**
This PR adds `sccache` and builds many of the project dependencies in the base image to avoid downloading and compiling them each time the project builds in CI.

### CODE REVIEW

**Comment by kou** [comment]
I'm testing this.
Could you show how to confirm whether cache is used or not?

**Comment by andygrove** [comment]
Thanks @kou  ... my advice would be to run `docker-compose build rust` then `docker-compose run rust` before and after this PR and count how many crates get *downloaded* (not compiled). You should see a significant reduction after this PR.

**Comment by kou** [comment]
Thanks.
Without this change https://github.com/apache/arrow/commit/02d1e9736808d9a9624bef5577c880d8c165e853 takes 14m 44s https://github.com/apache/arrow/runs/252662504
With this change takes 22m 46s https://github.com/apache/arrow/pull/5587/checks?check_run_id=253006461
It seems that this is slower on GitHub Actions.

And I got all "Compiling"s:

```text
   Compiling libc v0.2.62
   Compiling cfg-if v0.1.10
   Compiling autocfg v0.1.6
   Compiling serde v1.0.101
   Compiling semver-parser v0.7.0
   Compiling lazy_static v1.4.0
   Compiling log v0.4.8
   Compiling rand_core v0.4.2
   Compiling ryu v1.0.0
   Compiling byteorder v1.3.2
   Compiling proc-macro2 v1.0.5
   Compiling unicode-xid v0.2.0
   Compiling memchr v2.2.1
   Compiling itoa v0.4.4
   Compiling bitflags v1.2.0
   Compiling syn v1.0.5
   Compiling indexmap v1.2.0
   Compiling same-file v1.0.5
   Compiling packed_simd v0.3.3
   Compiling smallvec v0.6.10
   Compiling arrayvec v0.4.11
   Compiling regex-syntax v0.6.12
   Compiling nodrop v0.1.13
   Compiling scopeguard v1.0.0
   Compiling unicode-width v0.1.6
   Compiling ansi_term v0.11.0
   Compiling vec_map v0.8.1
   Compiling strsim v0.8.0
   Compiling rayon-core v1.6.0
   Compiling either v1.5.3
   Compiling version_check v0.1.5
   Compiling rustc-demangle v0.1.16
   Compiling cast v0.2.2
   Compiling pulldown-cmark v0.2.0
   Compiling void v1.0.2
   Compiling remove_dir_all v0.5.2
   Compiling glob v0.3.0
   Compiling bytecount v0.4.0
   Compiling glob v0.2.11
   Compiling crc32fast v1.2.0
   Compiling alloc-no-stdlib v1.3.0
   Compiling adler32 v1.0.4
   Compiling integer-encoding v1.0.7
   Compiling byteorder v0.5.3
   Compiling try_from v0.2.2
   Compiling parquet v1.0.0-SNAPSHOT (/arrow/rust/parquet)
   Compiling quick-error v1.2.2
   Compiling nix v0.13.1
   Compiling uuid v0.7.4
   Compiling encode_unicode v0.3.6
   Compiling unicode-segmentation v1.3.0
   Compiling utf8parse v0.1.1
   Compiling fnv v1.0.6
   Compiling thread_local v0.3.6
   Compiling crossbeam-utils v0.6.6
   Compiling rand_core v0.3.1
   Compiling rand_jitter v0.1.4
   Compiling num-traits v0.2.8
   Compiling num-integer v0.1.41
   Compiling num-bigint v0.2.3
   Compiling num-rational v0.2.2
   Compiling rand_pcg v0.1.2
   Compiling rand_chacha v0.1.1
   Compiling num-iter v0.1.39
   Compiling num-complex v0.2.3
   Compiling rand v0.6.5
   Compiling walkdir v2.2.9
   Compiling textwrap v0.11.0
   Compiling flatbuffers v0.5.0
   Compiling itertools v0.8.0
   Compiling unreachable v1.0.0
   Compiling error-chain v0.12.1
   Compiling brotli-decompressor v1.3.1
   Compiling miniz_oxide v0.3.3
   Compiling rand_xorshift v0.1.1
   Compiling rand_isaac v0.1.1
   Compiling rand_hc v0.1.0
   Compiling crossbeam-queue v0.1.2
   Compiling crossbeam-channel v0.3.9
   Compiling log v0.3.9
   Compiling regex-automata v0.1.8
   Compiling rand_xoshiro v0.1.0
   Compiling snap v0.2.5
   Compiling criterion-plot v0.3.1
   Compiling num_cpus v1.10.1
   Compiling jobserver v0.1.17
   Compiling rand_os v0.1.3
   Compiling time v0.1.42
   Compiling atty v0.2.13
   Compiling rand v0.4.6
   Compiling flate2 v1.0.12
   Compiling dirs v1.0.5
   Compiling rand v0.5.6
   Compiling quote v1.0.2
   Compiling brotli v2.5.1
   Compiling clap v2.33.0
   Compiling threadpool v1.7.1
   Compiling num-traits v0.1.43
   Compiling csv-core v0.1.6
   Compiling aho-corasick v0.7.6
   Compiling cc v1.0.45
   Compiling term v0.5.2
   Compiling thrift v0.0.4
   Compiling ordered-float v0.5.2
   Compiling tempdir v0.3.7
   Compiling chrono v0.4.9
   Compiling backtrace-sys v0.1.31
   Compiling zstd-sys v1.4.13+zstd.1.4.3
   Compiling lz4-sys v1.8.3
   Compiling regex v1.3.1
   Compiling parquet-format v2.5.0
   Compiling sqlparser v0.2.4
   Compiling semver v0.9.0
   Compiling serde_json v1.0.41
   Compiling bstr v0.2.8
   Compiling rustyline v4.1.0
   Compiling rustc_version v0.2.3
   Compiling serde_derive v1.0.101
   Compiling num v0.2.0
   Compiling memoffset v0.5.1
   Compiling backtrace v0.3.38
   Compiling crossbeam-epoch v0.7.2
   Compiling csv v1.1.1
   Compiling tinytemplate v1.0.2
   Compiling crossbeam-deque v0.7.1
   Compiling crossbeam v0.7.2
   Compiling prettytable-rs v0.8.0
   Compiling rayon v1.2.0
   Compiling cargo_metadata v0.6.4
   Compiling criterion v0.2.11
   Compiling skeptic v0.13.4
   Compiling lz4 v1.23.1
   Compiling zstd-safe v1.4.13+zstd.1.4.3
   Compiling zstd v0.4.28+zstd.1.4.3
   Compiling arrow v1.0.0-SNAPSHOT (/arrow/rust/arrow)
   Compiling datafusion v1.0.0-SNAPSHOT (/arrow/rust/datafusion)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 02s
    Finished dev [unoptimized + debuginfo] target(s) in 0.07s
     Running target/debug/deps/arrow-73bae392adbe9091
```

This doesn't work as expected?

**Comment by andygrove** [comment]
@kou Are you stating timing just for "run" or for "build" and "run" ? Overall this will take longer, but the "run" part should have been faster. This is only useful if the build step gets cached.

**Comment by kou** [comment]
> Are you stating timing just for "run" or for "build" and "run" ?

For GitHub Actions: "build" and "run".

For local confirmation: just "run".

I always get "Compiling" even when I run "run" multiple times:

```console
$ docker-compose run rust
ARNING: The CI_ARROW_SHA variable is not set. Defaulting to a blank string.
WARNING: The CI_ARROW_BRANCH variable is not set. Defaulting to a blank string.
Default host: x86_64-unknown-linux-gnu
rustup home:  /usr/local/rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu
nightly-2019-09-25-x86_64-unknown-linux-gnu (default)
1.38.0-x86_64-unknown-linux-gnu

active toolchain
----------------

nightly-2019-09-25-x86_64-unknown-linux-gnu (overridden by '/arrow/rust/rust-toolchain')
rustc 1.39.0-nightly (6ef275e6c 2019-09-24)

Running formatting checks ...
Formatting checks completed
   Compiling libc v0.2.62
   Compiling cfg-if v0.1.10
   Compiling autocfg v0.1.6
   Compiling serde v1.0.101
   Compiling lazy_static v1.4.0
   Compiling semver-parser v0.7.0
   Compiling log v0.4.8
   Compiling rand_core v0.4.2
   Compiling ryu v1.0.0
   Compiling byteorder v1.3.2
   Compiling proc-macro2 v1.0.5
   Compiling memchr v2.2.1
   Compiling itoa v0.4.4
   Compiling unicode-xid v0.2.0
   Compiling bitflags v1.2.0
   Compiling syn v1.0.5
   Compiling indexmap v1.2.0
   Compiling same-file v1.0.5
   Compiling packed_simd v0.3.3
   Compiling arrayvec v0.4.11
   Compiling smallvec v0.6.10
   Compiling regex-syntax v0.6.12
   Compiling nodrop v0.1.13
   Compiling scopeguard v1.0.0
   Compiling unicode-width v0.1.6
   Compiling vec_map v0.8.1
   Compiling ansi_term v0.11.0
   Compiling strsim v0.8.0
   Compiling version_check v0.1.5
   Compiling rayon-core v1.6.0
   Compiling either v1.5.3
   Compiling rustc-demangle v0.1.16
   Compiling cast v0.2.2
   Compiling pulldown-cmark v0.2.0
   Compiling glob v0.3.0
   Compiling void v1.0.2
   Compiling remove_dir_all v0.5.2
   Compiling bytecount v0.4.0
   Compiling glob v0.2.11
   Compiling crc32fast v1.2.0
   Compiling try_from v0.2.2
   Compiling alloc-no-stdlib v1.3.0
   Compiling adler32 v1.0.4
   Compiling integer-encoding v1.0.7
   Compiling byteorder v0.5.3
   Compiling parquet v1.0.0-SNAPSHOT (/arrow/rust/parquet)
   Compiling quick-error v1.2.2
   Compiling nix v0.13.1
   Compiling uuid v0.7.4
   Compiling encode_unicode v0.3.6
   Compiling utf8parse v0.1.1
   Compiling unicode-segmentation v1.3.0
   Compiling fnv v1.0.6
   Compiling thread_local v0.3.6
   Compiling crossbeam-utils v0.6.6
   Compiling rand_core v0.3.1
   Compiling rand_jitter v0.1.4
   Compiling num-traits v0.2.8
   Compiling num-integer v0.1.41
   Compiling num-bigint v0.2.3
   Compiling num-complex v0.2.3
   Compiling num-iter v0.1.39
   Compiling rand_chacha v0.1.1
   Compiling rand_pcg v0.1.2
   Compiling num-rational v0.2.2
   Compiling rand v0.6.5
   Compiling walkdir v2.2.9
   Compiling textwrap v0.11.0
   Compiling flatbuffers v0.5.0
   Compiling itertools v0.8.0
   Compiling error-chain v0.12.1
   Compiling unreachable v1.0.0
   Compiling miniz_oxide v0.3.3
   Compiling brotli-decompressor v1.3.1
   Compiling rand_hc v0.1.0
   Compiling rand_xorshift v0.1.1
   Compiling rand_isaac v0.1.1
   Compiling crossbeam-queue v0.1.2
   Compiling crossbeam-channel v0.3.9
   Compiling log v0.3.9
   Compiling regex-automata v0.1.8
   Compiling rand_xoshiro v0.1.0
   Compiling snap v0.2.5
   Compiling criterion-plot v0.3.1
   Compiling num_cpus v1.10.1
   Compiling jobserver v0.1.17
   Compiling time v0.1.42
   Compiling rand_os v0.1.3
   Compiling atty v0.2.13
   Compiling rand v0.4.6
   Compiling flate2 v1.0.12
   Compiling dirs v1.0.5
   Compiling rand v0.5.6
   Compiling quote v1.0.2
   Compiling brotli v2.5.1
   Compiling threadpool v1.7.1
   Compiling num-traits v0.1.43
   Compiling clap v2.33.0
   Compiling csv-core v0.1.6
   Compiling aho-corasick v0.7.6
   Compiling term v0.5.2
   Compiling cc v1.0.45
   Compiling ordered-float v0.5.2
   Compiling thrift v0.0.4
   Compiling chrono v0.4.9
   Compiling tempdir v0.3.7
   Compiling backtrace-sys v0.1.31
   Compiling zstd-sys v1.4.13+zstd.1.4.3
   Compiling lz4-sys v1.8.3
   Compiling parquet-format v2.5.0
   Compiling regex v1.3.1
   Compiling sqlparser v0.2.4
   Compiling rustyline v4.1.0
   Compiling semver v0.9.0
   Compiling serde_json v1.0.41
   Compiling bstr v0.2.8
   Compiling rustc_version v0.2.3
   Compiling serde_derive v1.0.101
   Compiling num v0.2.0
   Compiling memoffset v0.5.1
   Compiling crossbeam-epoch v0.7.2
   Compiling backtrace v0.3.38
   Compiling csv v1.1.1
   Compiling tinytemplate v1.0.2
   Compiling crossbeam-deque v0.7.1
   Compiling crossbeam v0.7.2
   Compiling prettytable-rs v0.8.0
   Compiling rayon v1.2.0
   Compiling cargo_metadata v0.6.4
   Compiling criterion v0.2.11
   Compiling skeptic v0.13.4
   Compiling lz4 v1.23.1
   Compiling zstd-safe v1.4.13+zstd.1.4.3
   Compiling zstd v0.4.28+zstd.1.4.3
   Compiling arrow v1.0.0-SNAPSHOT (/arrow/rust/arrow)
   Compiling datafusion v1.0.0-SNAPSHOT (/arrow/rust/datafusion)
    Finished dev [unoptimized + debuginfo] target(s) in 59.78s
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running target/debug/deps/arrow-73bae392adbe9091
...
$ docker-compose run rust
WARNING: The CI_ARROW_SHA variable is not set. Defaulting to a blank string.
WARNING: The CI_ARROW_BRANCH variable is not set. Defaulting to a blank string.
Default host: x86_64-unknown-linux-gnu
rustup home:  /usr/local/rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu
nightly-2019-09-25-x86_64-unknown-linux-gnu (default)
1.38.0-x86_64-unknown-linux-gnu

active toolchain
----------------

nightly-2019-09-25-x86_64-unknown-linux-gnu (overridden by '/arrow/rust/rust-toolchain')
rustc 1.39.0-nightly (6ef275e6c 2019-09-24)

Running formatting checks ...
Formatting checks completed
   Compiling libc v0.2.62
   Compiling cfg-if v0.1.10
   Compiling autocfg v0.1.6
   Compiling serde v1.0.101
   Compiling lazy_static v1.4.0
   Compiling semver-parser v0.7.0
   Compiling log v0.4.8
   Compiling ryu v1.0.0
   Compiling rand_core v0.4.2
   Compiling byteorder v1.3.2
   Compiling proc-macro2 v1.0.5
   Compiling itoa v0.4.4
   Compiling bitflags v1.2.0
   Compiling unicode-xid v0.2.0
   Compiling memchr v2.2.1
   Compiling syn v1.0.5
   Compiling indexmap v1.2.0
   Compiling same-file v1.0.5
   Compiling packed_simd v0.3.3
   Compiling smallvec v0.6.10
   Compiling arrayvec v0.4.11
   Compiling regex-syntax v0.6.12
   Compiling nodrop v0.1.13
   Compiling scopeguard v1.0.0
   Compiling unicode-width v0.1.6
   Compiling vec_map v0.8.1
   Compiling strsim v0.8.0
   Compiling ansi_term v0.11.0
   Compiling rayon-core v1.6.0
   Compiling either v1.5.3
   Compiling version_check v0.1.5
   Compiling rustc-demangle v0.1.16
   Compiling cast v0.2.2
   Compiling pulldown-cmark v0.2.0
   Compiling glob v0.3.0
   Compiling remove_dir_all v0.5.2
   Compiling void v1.0.2
   Compiling glob v0.2.11
   Compiling bytecount v0.4.0
   Compiling crc32fast v1.2.0
   Compiling alloc-no-stdlib v1.3.0
   Compiling integer-encoding v1.0.7
   Compiling byteorder v0.5.3
   Compiling adler32 v1.0.4
   Compiling try_from v0.2.2
   Compiling parquet v1.0.0-SNAPSHOT (/arrow/rust/parquet)
   Compiling quick-error v1.2.2
   Compiling nix v0.13.1
   Compiling utf8parse v0.1.1
   Compiling encode_unicode v0.3.6
   Compiling unicode-segmentation v1.3.0
   Compiling uuid v0.7.4
   Compiling fnv v1.0.6
   Compiling crossbeam-utils v0.6.6
   Compiling thread_local v0.3.6
   Compiling rand_core v0.3.1
   Compiling rand_jitter v0.1.4
   Compiling num-traits v0.2.8
   Compiling num-integer v0.1.41
   Compiling num-bigint v0.2.3
   Compiling rand_pcg v0.1.2
   Compiling num-rational v0.2.2
   Compiling num-iter v0.1.39
   Compiling num-complex v0.2.3
   Compiling rand_chacha v0.1.1
   Compiling rand v0.6.5
   Compiling walkdir v2.2.9
   Compiling textwrap v0.11.0
   Compiling flatbuffers v0.5.0
   Compiling itertools v0.8.0
   Compiling error-chain v0.12.1
   Compiling unreachable v1.0.0
   Compiling brotli-decompressor v1.3.1
   Compiling miniz_oxide v0.3.3
   Compiling rand_isaac v0.1.1
   Compiling rand_xorshift v0.1.1
   Compiling rand_hc v0.1.0
   Compiling crossbeam-queue v0.1.2
   Compiling crossbeam-channel v0.3.9
   Compiling log v0.3.9
   Compiling regex-automata v0.1.8
   Compiling rand_xoshiro v0.1.0
   Compiling snap v0.2.5
   Compiling criterion-plot v0.3.1
   Compiling quote v1.0.2
   Compiling num_cpus v1.10.1
   Compiling jobserver v0.1.17
   Compiling rand_os v0.1.3
   Compiling time v0.1.42
   Compiling atty v0.2.13
   Compiling rand v0.4.6
   Compiling flate2 v1.0.12
   Compiling dirs v1.0.5
   Compiling rand v0.5.6
   Compiling brotli v2.5.1
   Compiling threadpool v1.7.1
   Compiling clap v2.33.0
   Compiling num-traits v0.1.43
   Compiling csv-core v0.1.6
   Compiling aho-corasick v0.7.6
   Compiling term v0.5.2
   Compiling cc v1.0.45
   Compiling thrift v0.0.4
   Compiling ordered-float v0.5.2
   Compiling chrono v0.4.9
   Compiling tempdir v0.3.7
   Compiling regex v1.3.1
   Compiling backtrace-sys v0.1.31
   Compiling zstd-sys v1.4.13+zstd.1.4.3
   Compiling lz4-sys v1.8.3
   Compiling parquet-format v2.5.0
   Compiling semver v0.9.0
   Compiling serde_json v1.0.41
   Compiling bstr v0.2.8
   Compiling serde_derive v1.0.101
   Compiling rustyline v4.1.0
   Compiling sqlparser v0.2.4
   Compiling rustc_version v0.2.3
   Compiling num v0.2.0
   Compiling csv v1.1.1
   Compiling tinytemplate v1.0.2
   Compiling memoffset v0.5.1
   Compiling crossbeam-epoch v0.7.2
   Compiling backtrace v0.3.38
   Compiling prettytable-rs v0.8.0
   Compiling crossbeam-deque v0.7.1
   Compiling crossbeam v0.7.2
   Compiling rayon v1.2.0
   Compiling cargo_metadata v0.6.4
   Compiling skeptic v0.13.4
   Compiling criterion v0.2.11
   Compiling lz4 v1.23.1
   Compiling zstd-safe v1.4.13+zstd.1.4.3
   Compiling zstd v0.4.28+zstd.1.4.3
   Compiling arrow v1.0.0-SNAPSHOT (/arrow/rust/arrow)
   Compiling datafusion v1.0.0-SNAPSHOT (/arrow/rust/datafusion)
    Finished dev [unoptimized + debuginfo] target(s) in 57.95s
    Finished dev [unoptimized + debuginfo] target(s) in 0.06s
     Running target/debug/deps/arrow-73bae392adbe9091
...
```

File: rust/Cargo.toml.cache [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+[package]
+name = "arrow-dependencies"
+version = "0.0.0"
+
+[dependencies]
+brotli = "2.5"
```
Comment (nevi-me) [inline rust/Cargo.toml.cache]: Will this mean that when we add a new dependency or change a version, we should change this too?

Comment (andygrove) [reply rust/Cargo.toml.cache]: Yes, ideally, to maintain the benefit of caching dependencies.

**Review by andygrove — COMMENTED** [review]
I'm tempted to close this PR since there do not seem to be clear benefits of doing this.

**Comment by andygrove** [comment]
This is complex, and doesn't seem to be helping IRL, so I'm closing this. Thanks for testing it @kou

**Pull request closed without merging**